### DISCUSSION TITLE
COG, crop, projections on Google shifted North a bit...

### DISCUSSION THREAD

**Discussion by alturic** [developmentseed/titiler#564] [Q&A]
Hi, we have been using TiTiler for XYZ tiles for a while now, and as we have an immense amount of tiles we are running into 20+ second wait times for all of the tiles to be retrieved as we are hitting the browser connection limits (even splitting the tiles across multiple domains) so we wanted to try using the /crop endpoints and it seemingly works extremely well with one small issue... the image is shifted North in relation to where it should actually be placed.

We are using existing code that worked flawlessly to display the proper sized single image PNG's so I'm struggling to find out where the problem actually lies, or whether /crop would actually allow us to do this, even though it definitely seems like it works as we're expecting with the exception fo the image being shifted just a tad, which is why I'm thinking it's a projection issue?

**Comment by vincentsarago** [comment]
@alturic 🤷 I can't help without a minimum reproductible example

**Reply by alturic** [reply]
I understand. With all of the non-generic Google code we have that would be really difficult sadly.

Nothing stands out if the tiles endpoint returns the same exact COG perfectly, but the crop endpoint doesn't seem to be? I admit I'm a greenhorn when it comes to the "how" and "what" TiTiler is ultimately doing when it comes to chopping up these images. I'm unsure if it's something related to TiTiler or if it's something related to our Google code since the /crop PNG's are using OverlayView() function of GMaps to put them into a div.

I just didn't know if something stood out to where /tiles uses the projection of the COG and /crop wouldn't or something because it's just using a bounding box with a width and height.

Perhaps I should update TiTiler just to confirm things as well.

**Reply by vincentsarago** [reply]
FYI `rio_tiler.io.Reader.tile()` uses `rio_tiler.io.Reader.part()` which is used in the `/crop` endpoint https://github.com/developmentseed/titiler/blob/master/src/titiler/core/titiler/core/factory.py#L937-L941

https://github.com/cogeotiff/rio-tiler/blob/bcd57f8d5257eaed554f108b4ab920cf3c7ff67a/rio_tiler/io/rasterio.py#L355-L366

the issue might be in the frontend or the coordinates and CRS you use 🤷

**Reply by alturic** [reply]
Yea, I was thinking it *might* be something with our GMaps code as it's the same underlying COG that the tile endpoint is returning for the tiles, which is why I was scratching my head.

**Reply by alturic** [reply]
@vincentsarago Actually, let me ask you this, does /crop reproject to 3857 by default or does it just return a flat image?

**Reply by vincentsarago** [reply]
@alturic by default the output projection is the same as the input projection WGS84 (epss:4326)

https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/io/rasterio.py#L372

**Reply by alturic** [reply]
Vincent, I’m going to work on getting some code, but unless 0.7.0 doesn't use the same output logic for /crop as it does /tile, you would expect a geotiff that produces tiles perfect to also be projected correctly from the /crop endpoint, no?

Would a reproducible example be “enough” to show you on the Google USGS Overlay example? That’s ultimately what I’m looking to do with the /crop endpoint… take a large cropped image and put it on the map. I know I’m definitely pulling in the correct lat/lng’s as it’s straight from Google Maps API, but even using their basic code it’s not coming out as expected. It seems like the /crop endpoint is “squishing” the real bounds which when projected on Google it’s being shifted north. Oddly enough the further you zoom in the less the shift is noticeable.

Let me work on getting something to you.

**Reply by alturic** [reply]
Ok, here is what I came up with, I'm unsure if this is "enough". https://jsfiddle.net/m6pvLe0d/1/

Now, the cut-off on the left side of the United States is expected. This is the exact lat/lon I'm expecting the /crop image to be. It's 1903x854 resolution as that's my map div's resolution. The easiest way to tell something is wrong, is by looking at the Great Lakes in the US. You can clearly see the "definition" of them on the colored image, and then Toggle the image on/off and you can see they don't match up.

As I said in my most recent reply, and again I'm fairly new to the mapping world, I kind've just assumed the same underlying image that's generated the below (and generates tiles perfectly), would just work when using /crop.

```
gdal_translate -co "TILED=YES" -of Gtiff -a_ullr -128 58 -65 20 -a_srs "EPSG:4326" image.png image_geotiff.tiff

rio cogeo create -w --use-cog-driver image_geotiff.tiff rio_image_geotiff.tiff
```

And yes, this two above commands are used to generate the underlying geotiff you see in the fiddle, with tiles being correctly made/projected, but that cropped image not.

**Reply by alturic** [reply]
@vincentsarago Did you have any time to look at the fiddle? I also did a rio info on the source tiff (the one that's making tiles correctly, but cropping it incorrectly/shifting it north.

```
Driver: GTiff
File: /store/current-radar/202212201850.tiff
COG: True
Compression: DEFLATE
ColorSpace: None

Profile
    Width:            11776
    Height:           9728
    Bands:            4
    Tiled:            True
    Dtype:            uint8
    NoData:           None
    Alpha Band:       True
    Internal Mask:    False
    Interleave:       PIXEL
    ColorMap:         False
    ColorInterp:      ('red', 'green', 'blue', 'alpha')
    Scales:           (1.0, 1.0, 1.0, 1.0)
    Offsets:          (0.0, 0.0, 0.0, 0.0)

Geo
    Crs:              EPSG:3857
    Origin:           (-14401959.12137977, 8140237.7642581295)
    Resolution:       (611.49622628141, -611.49622628141)
    BoundingBox:      (-14401959.12137977, 2191602.4749925723, -7200979.560689886, 8140237.7642581295)
    MinZoom:          2
    MaxZoom:          8

Image Metadata
    AREA_OR_POINT: Area
    OVR_RESAMPLING_ALG: NEAREST

Image Structure
    COMPRESSION: DEFLATE
    INTERLEAVE: PIXEL
    LAYOUT: COG

Tiling Scheme
    NAME: GOOGLEMAPSCOMPATIBLE
    ZOOM_LEVEL: 7

Band 1
    ColorInterp: red

Band 2
    ColorInterp: green

Band 3
    ColorInterp: blue

Band 4
    ColorInterp: alpha

IFD
    Id      Size           BlockSize     Decimation
    0       11776x9728     512x512       0
    1       5888x4864      512x512       2
    2       2944x2432      512x512       4
    3       1472x1216      512x512       8
    4       736x608        512x512       16
    5       368x304        512x512       32
```

**Reply by vincentsarago** [reply]
@alturic sorry I didn't had the time to check yet!

**Comment by alturic** [comment]
@vincentsarago I believe that in 0.7 the /crop endpoint doesn’t return the image (png in my case) georeferenced even if the originating GTiff is properly referenced, can you confirm of that is the case? By looking at the fiddle I am almost certain that is the issue.

I would update to 0.10 but there is BC’s that I can’t easily fix being so new to Python so I just want to confirm the georeference issue first before I go breaking things while trying to fix those BCs.

**Comment by vincentsarago** [comment]
@alturic I can't reproduce the issue 
```
$ uvicorn titiler.application.main:app --port 8000 --reload

$ curl http://127.0.0.1:8000/cog/crop/-10,-10,10,10.tif?url=.github/data/world.tif > crop.tif

## Check the geo information of the tif
$ rio cogeo info crop.tiff 
Driver: GTiff
File: /Users/vincentsarago/Desktop/crop.tiff
COG: False
Compression: None
ColorSpace: None

Profile
    Width:            1024
    Height:           1024
    Bands:            4
    Tiled:            False
    Dtype:            uint8
    NoData:           None
    Alpha Band:       True
    Internal Mask:    False
    Interleave:       PIXEL
    ColorMap:         False
    ColorInterp:      ('red', 'green', 'blue', 'alpha')
    Scales:           (1.0, 1.0, 1.0, 1.0)
    Offsets:          (0.0, 0.0, 0.0, 0.0)

Geo
    Crs:              EPSG:4326
    Origin:           (-10.0, 10.0)
    Resolution:       (0.01953125, -0.01953125)
    BoundingBox:      (-10.0, -10.0, 10.0, 10.0)
    MinZoom:          4
    MaxZoom:          6

# Then load the tif in QGIS with `.github/data/world.tif` in background
```

![2023-01-05 12-36-24 2023-01-05 12_36_50](https://user-images.githubusercontent.com/10407788/210771975-e97f9054-ffc6-4322-8edc-1fdf1780fdf9.gif)

##### With GOES input file (in GEOS projection)
![2023-01-05 12-43-20 2023-01-05 12_43_41](https://user-images.githubusercontent.com/10407788/210772772-cce54a4e-d511-4dad-9c19-8718da0c0f60.gif)

**Reply by alturic** [reply]
So I fixed the problem. I don't know "what" it actually "is" as again I'm fresh with mapping and georeferencing images, but I'm on 0.7 TiTiler, and 3.1.6 riotiler, I ultimately did this in cogeo.py (which I know is not proper):

```
    def part(
        self,
        bbox: BBox,
        #dst_crs: Optional[CRS] = None,
        dst_crs: Optional[CRS] = 'EPSG:3857',
        bounds_crs: CRS = WGS84_CRS,
        indexes: Optional[Union[int, Sequence]] = None,
        expression: Optional[str] = None,
        max_size: Optional[int] = None,
        height: Optional[int] = None,
        width: Optional[int] = None,
        **kwargs: Any,
    ) -> ImageData:
```

And the output PNG matches up perfectly. What's interesting to me, regardless if I used 4326 or 3857 as the input COG, both produced the same unaligned output prior to doing that.

**Reply by vincentsarago** [reply]
@alturic https://github.com/developmentseed/titiler/discussions/564#discussioncomment-4397174

as I mentioned in ☝️ by default titiler/rio-tiler will assume the bounds CRS to be in EPSG:4326 and if `dst_crs` is not specified, it will return data in the same CRS as the bounds.... 

The dataset CRS has no influence on the output CRS.

**Reply by vincentsarago** [reply]
you can easily create TiTiler endpoints with dst_crs set to "epsg:3857"

```python
from typing import Optional
from titiler.core.dependencies import DefaultDependency

from dataclasses import dataclass
from fastapi import Query

@dataclass
class ImageParams(DefaultDependency):
    """Common Preview/Crop parameters."""

    max_size: Optional[int] = Query(
        1024, description="Maximum image size to read onto."
    )
    height: Optional[int] = Query(None, description="Force output image height.")
    width: Optional[int] = Query(None, description="Force output image width.")
    dst_crs: Optional[str] = Query("espg:3857", description="Output Image CRS.")

    def __post_init__(self):
        """Post Init."""
        if self.width and self.height:
            self.max_size = None


cog = TileFactory(img_dependency=ImageParams)
```
